src/java/org/apache/cassandra/cql3/Operator.java

```diff
@@ -324,23 +561,150 @@ public static Operator readFrom(DataInput input) throws IOException
+            case MAP_ELEMENT:
+                ColumnMetadata column = expression.firstColumn();
+                AbstractType<?> type = column.type;
+                if (type.isMultiCell())
+                {
+                    // Non-frozen UDTs don't support any operator
+                    checkFalse(type.isUDT(),
+                               "Non-frozen UDT column '%s' (%s) cannot be restricted by any relation",
+                               column.name,
+                               type.asCQL3Type());
+
+                    // We don't support relations against entire collections (unless they're frozen), like "numbers = {1, 2, 3}"
+                    checkFalse(type.isCollection()
+                                    && !this.appliesToMapKeys()
+                                    && !this.appliesToCollectionElements()
+                                    && expression.kind() != ColumnsExpression.Kind.MAP_ELEMENT,
+                               "Collection column '%s' (%s) cannot be restricted by a '%s' relation",
+                               column.name,
+                               type.asCQL3Type(),
+                               this);
+                }
+                break;
+        }
+    }
+
     /**
-     * Checks if this operator is a slice operator.
+     * Checks if the specified expression kind can be used with this operator.
+     * @param kind the expression kind
+     * @return {@code true} if the specified expression kind can be used with this operator, {@code false} otherwise.
+     */
+    public boolean canBeUsedWith(ColumnsExpression.Kind kind)
+    {
+        // All operators support single columns
+        return kind == ColumnsExpression.Kind.SINGLE_COLUMN;
+    }
+
+    /**
+     * Checks if this operator applies to non-multicell column values.
+     * @return {@code true} if this operator applies to column values, {@code false} otherwise.
+     */
+    public boolean appliesToColumnValues()
+    {
+        return true;
+    }
+
+    /**
+     * Checks if this operator applies to collection elements (from frozen and non-frozen collections).
+     * @return {@code true} if this operator applies to collection elements, {@code false} otherwise.
+     */
+    public boolean appliesToCollectionElements()
+    {
+        return false;
+    }
+
+    /**
+     * Checks if this operator applies to map keys.
+     * @return {@code true} if this operator applies to map keys, {@code false} otherwise.
+     */
+    public boolean appliesToMapKeys()
+    {
+        return false;
+    }
+
+    /**
+     * Restricts the specified range set based on the operator arguments (optional operation).
+     * @param rangeSet the range set to restrict
+     * @param args the operator arguments
+     * @return the restricted range set
+     */
+    public RangeSet<ClusteringElements> restrict(RangeSet<ClusteringElements> rangeSet, List<ClusteringElements> args)
+    {
+        throw new UnsupportedOperationException();
```

Comment:
```suggestion
        throw new UnsupportedOperationException("Cannot restrict " + this + " operator");
```
...or something similar?